server/store/service.go

```diff
@@ -49,24 +50,63 @@ func NewService(mm *pluginapi.Client, conf config.Service) *Service {
 	return s
 }
 
-func (s *Service) hashkey(globalPrefix, namespace, prefix, id string) string {
-	if id == "" || namespace == "" {
-		return ""
+func (s *Service) hashkey(globalNamespace, botUserID, appNamespace, key string) (string, error) {
+	gns := []byte(globalNamespace)
+	b := []byte(botUserID)
+	k := []byte(key)
+
+	ns := []byte(appNamespace)
+	switch len(ns) {
+	case 0:
+		ns = []byte{' ', ' '}
+	case 1:
+		ns = []byte{ns[0], ' '}
+	case 2:
+		// nothing to do
+	default:
+		return "", errors.Errorf("prefix %q is longer than the limit of 2 ASCII characters", appNamespace)
+	}
+
+	switch {
+	case len(k) == 0:
+		return "", errors.New("key must not be empty")
+	case len(b) != 26:
+		return "", errors.Errorf("botUserID %q must be exactly 26 ASCII characters", botUserID)
+	case len(gns) != 2 || gns[0] != '.':
+		return "", errors.Errorf("global prefix %q is not 2 ASCII characters starting with a '.'", globalNamespace)
+	}
+
+	hashed := hashkey(gns, b, ns, k)
+	if len(hashed) > model.KEY_VALUE_KEY_MAX_RUNES {
```

Comment:
Nit: This seems like a thing that should be covered by tests, not checked at runtime. If `hashkey` is working correctly, the length is never more than `model.KEY_VALUE_KEY_MAX_RUNES`.